mypy/meet.py

```diff
@@ -77,6 +78,10 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
           and narrowed.type.is_metaclass()):
         # We'd need intersection types, so give up.
         return declared
+    elif isinstance(narrowed, TypeVarType):
+        if is_same_type(narrowed.upper_bound, meet_types(narrowed.upper_bound, declared)):
```

Comment:
Does it work to check `is_subtype(narrowed.upper_bound, declared)` instead?

Mathematically, the two should be equivalent: x ≤ y if and only if x ∧ y = x (see https://en.wikipedia.org/wiki/Join_and_meet).

In practice, mypy's implementations fall short of mathematical perfection, and I think that `is_subtype` is more reliable than `meet_types`, and reads more naturally.

Reply:
The difference from `is_subtype` is how the `UnboundType` is treated. When type of `declared` is `UnboundType`, `is_subtype` will always return `True`. But since the type of `meet_type` result in this case is either `AnyType`, `NoneType` or `UninhabitedType`, `is_same_type` will result in `True` only if `narrowed.upper_bound` is also of `AnyType`, `NoneType`, `UninhabitedType` or `UnboundType`. I thought it is better not to unconditionally narrow unbound type with type variable.

Reply:
Hm. OK, that seems plausible. Do you have an example where we get different behavior between the two? I think that would make it easier for me to consider the difference.

Reply:
I tried to come up with an example and while I was skimming through `typeanal.py` to figure out how one can get an unbound type during type checking phase, I've learned about [this issue](https://github.com/python/mypy/issues/4987). Seems that there is no need to consider unbound types in the type checker, unless their presence leads to crashes. And since  `is_subtype` works fine with unbound types, I think it is actually a good idea to rewrite the entire PR code like this:

```python
elif isinstance(narrowed, TypeVarType) and is_subtype(narrowed.upper_bound, declared):
    return narrowed
```

Reply:
OK, I agree!

Reply:
Done.